cloudsync_onedrive.py

```diff
@@ -991,131 +847,127 @@ def create(self, path, file_like, metadata=None) -> 'OInfo':
         if size == 0:
             if self.exists_path(path):
                 raise CloudFileExistsError()
-            with self._api() as client:
-                api_path = self._get_item(client, oid=pid).api_path
-                base = base.replace("'", "''")
-                name = urllib.parse.quote(base)
-                api_path += "/children('" + name + "')/content"
-                try:
-                    headers = {'content-type': 'text/plain'}
-                    r = self._direct_api("put", api_path, data=file_like, headers=headers)  # default timeout ok, size == 0 from "if" condition
-                except CloudTemporaryError:
-                    info = self.info_path(path)
-                    # onedrive can fail with ConnectionResetByPeer, but still secretly succeed... just without returning info
-                    # if so, check hashes, and if all is OK, return OK
-                    if info and info.hash == self.hash_data(file_like):
-                        return info
-                    # alternately this could be a race condition, where two people upload at once
-                    # so fail otherwise
-                    raise
-            return self._info_from_rest(r, root=dirname)
+
+            base = base.replace("'", "''")
+            name = urllib.parse.quote(base)
+            api_path = self._api_path(oid=pid) + "/children('" + name + "')/content"
+            try:
+                headers = {'content-type': 'text/plain'}
+                r = self._direct_api("put", api_path, data=file_like, headers=headers)  # default timeout ok, size == 0 from "if" condition
+                return self._info_from_rest(r, root=dirname)
+            except CloudTemporaryError:
+                info = self.info_path(path)
+                # onedrive can fail with ConnectionResetByPeer, but still secretly succeed... just without returning info
+                # if so, check hashes, and if all is OK, return OK
+                if info and info.hash == self.hash_data(file_like):
+                    return info
+                # alternately this could be a race condition, where two people upload at once
+                # so fail otherwise
+                raise
+
         else:
-            with self._api() as client:
-                r = self._upload_large(self._get_item(client, path=path).api_path, file_like, conflict="fail")
+            api_path = self._api_path(path=path)
+            r = self._upload_large(api_path, file_like, conflict="fail")
             return self._info_from_rest(r, root=self.dirname(path))
 
     def _upload_large(self, drive_path, file_like, conflict):  # pylint: disable=too-many-locals
-        with self._api():
```

Comment:
I think it isn't OK to remove `with self._api():` here. This provides a mutex around a loop of partial uploads, rather than simply around each individual partial upload.

Reply:
i put it back.

don't think this helps all that much given that the lock is for the current process only, but it is something. it does prevent other threads from taking the mutex mid-operation, which is a good thing.